src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -29,21 +29,22 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
-            _factory.GetTempData(context.HttpContext).Save();
         }
 
         /// <inheritdoc />
         public void OnResultExecuting(ResultExecutingContext context)
         {
+            if (context.Result is IKeepTempDataResult)
+            {
+                _factory.GetTempData(context.HttpContext).Keep();
+            }
+
+            _factory.GetTempData(context.HttpContext).Save();
```

Comment:
This is a breaking change in behavior. Now anything that mutates temp data in an action result won't have an effect.